src/app/loadout/reducer.ts

```diff
@@ -0,0 +1,51 @@
+import { produce } from 'immer';
+import { Reducer } from 'redux';
+import { ActionType, getType } from 'typesafe-actions';
+import * as actions from './actions';
+import { Loadout } from './loadout-types';
+
+export interface LoadoutsState {
+  /** A stack of previous loadouts by character ID, for undo loadout. */
+  readonly previousLoadouts: { [characterId: string]: Loadout[] };
+  /** The currently selected store for the loadouts page. */
+  readonly loadoutStoreId: string | undefined;
+}
+
+export type LoadoutsAction = ActionType<typeof actions>;
+
+const initialState: LoadoutsState = {
+  previousLoadouts: {},
+  loadoutStoreId: undefined,
+};
+
+export const loadouts: Reducer<LoadoutsState, LoadoutsAction> = (
+  state: LoadoutsState = initialState,
+  action: LoadoutsAction,
+): LoadoutsState => {
+  switch (action.type) {
+    case getType(actions.savePreviousLoadout): {
+      return produce(state, (draft) => {
+        const { storeId, loadoutId, previousLoadout } = action.payload;
+        let previousLoadouts = state.previousLoadouts[storeId] || [];
+        const lastPreviousLoadout = previousLoadouts.at(-1);
+        previousLoadouts =
+          lastPreviousLoadout && loadoutId === lastPreviousLoadout.id
+            ? // Pop the previous loadout since we're reapplying it
+              previousLoadouts.filter((l) => l.id !== loadoutId)
+            : // Push the previous loadout
+              [...previousLoadouts, previousLoadout];
+
+        draft.previousLoadouts[storeId] = previousLoadouts;
```

Comment:
This is the same as previous, I just moved it inside produce and rather than spreading at the end we have this assignment.

Reply:
I'm not sure there's a lot of benefit to doing that - I tend to try to avoid immer unless the mutation is really complicated. In this case we're already handling the complex list changes.

Reply:
Ah, last I remember we defaulted to it. Makes sense though, can revert.

Reply:
Done